File: strategy/marvin/task/victory.lua
local SuggestPass = require "task/ability/suggestpass"
local Victory = Class("Task.Victory", require "task/base", SuggestPass)

local ToTarget = require "trajectory/totarget"
local geom = require "../base/geom"

local NUM_OF_REVOLUTIONS = 2

function Victory:_init(center, startingAngle, angle, radius)
	assert(center and angle, "Missing Parameters for Victory-Task")
	self._center = center
	self._centerAngle = startingAngle
	self._radius = radius

	self._state = "double circle"
	self._outerAngle = angle
	self._ticks = 1
	self._increment = true
end

function Victory:run()
	self._centerAngle = self._centerAngle + math.pi / 480
	self._outerAngle = self._outerAngle + math.pi / (180 + self._ticks*180)
	local pos, dir
	if self._state == "double circle" then
		local origin = Vector.fromAngle(self._centerAngle):setLength(self._radius / 2)
		pos = self._center + origin + Vector.fromAngle(self._outerAngle):setLength(self._radius / (4 - self._ticks*2))
		dir = (self._center + origin - pos):angle()
		self._state = math.abs(self._centerAngle) < NUM_OF_REVOLUTIONS * 2 * math.pi and "double circle" or "spiral prepare"
		if self._state == "spiral prepare" then
			self._ticks = 1
			self._increment = false
		end
	elseif self._state == "spiral prepare" then
		local origin = Vector.fromAngle(self._centerAngle):setLength((self._radius / 2) * self._ticks)
		pos = self._center + origin + Vector.fromAngle(self._outerAngle):setLength(self._radius / (4 - self._ticks*2))
		dir = (self._center + origin - pos):angle()
		self._state = self._ticks > 0 and "spiral prepare" or "spiral"
	elseif self._state == "spiral" then
		pos = self._center + Vector.fromAngle(self._outerAngle):setLength(self._radius - self._ticks * self._radius * 3/4)
		dir = (self._center - self._robot.pos):angle()
		self._state = self._centerAngle > 2 * NUM_OF_REVOLUTIONS * 2 * math.pi and "double circle prepare" or "spiral"
		if self._state == "double circle prepare" then
			self._centerAngle = geom.normalizeAnglePositive(self._centerAngle)
			self._ticks = 0
			self._increment = true
		end
	elseif self._state == "double circle prepare" then
		local origin = Vector.fromAngle(self._centerAngle):setLength((self._radius / 2) * self._ticks)
		pos = self._center + origin + Vector.fromAngle(self._outerAngle):setLength(self._radius / (4 - self._ticks*2))
		dir = (self._center + origin - pos):angle()
		self._state = self._ticks > 1 and "double circle" or "double circle prepare"
	end

	if self._increment then
		self._ticks = self._ticks + 0.002
		self._increment = self._ticks < 1
	else
		self._ticks = self._ticks - 0.002
		self._increment = self._ticks < 0
	end
	self._robot.path:clearObstacles()
	local endSpeed = Vector(0, 0)
	self._robot.trajectory:update(ToTarget, pos, dir, 1, endSpeed)
end

return Victory

File: strategy/typescript/glados/task/test/victory.ts
import * as geom from "base/geom";
import { Position, Vector } from "base/vector";

import { Behavior } from "glados/agent/base/behavior";
import { Task } from "glados/task/base";
import * as PathHelper from "glados/trajectory/pathhelper";
import { ToTarget } from "glados/trajectory/totarget";

const NUM_OF_REVOLUTIONS = 3;
const ANGULAR_SPEED_FACTOR = 0.8; // the higher it is, the longer it takes

export class Victory extends Task {

	private _center: Position;
	private _centerAngle: number;
	private _radius: number;

	private _state: string;
	private _outerAngle: number;
	private _ticks: number;
	private _increment: boolean;

	private _obstacleTable: PathHelper.PathHelperParameters;

	public constructor(behavior: Behavior, center: Position, startingAngle: number, angle: number, radius: number) {
		if (center == undefined || angle == undefined) {
			throw new Error("Missing Parameters for Victory-Task");
		}
		super(behavior);
		this._center = center;
		this._centerAngle = startingAngle;
		this._radius = radius;

		this._state = "double circle";
		this._outerAngle = angle;
		this._ticks = 1;
		this._increment = true;

		let ignore = true;
		this._obstacleTable = {
			ignoreBall: ignore,
			ignorePass: ignore,
			ignoreGoals: ignore,
			ignoreDefenseArea: ignore
		};
	}

	public run() {
		this._centerAngle = this._centerAngle + Math.PI / (480 * ANGULAR_SPEED_FACTOR);
		this._outerAngle = this._outerAngle + Math.PI / (180 + this._ticks * 180) * ANGULAR_SPEED_FACTOR;
		let pos: Position | undefined;
		if (this._state === "double circle") {
			let origin = Vector.fromPolar(this._centerAngle, this._radius / 2);
			pos = this._center + origin + Vector.fromPolar(this._outerAngle, this._radius / (4 - this._ticks * 2));
			this._state = Math.abs(this._centerAngle) < NUM_OF_REVOLUTIONS * 2 * Math.PI ? "double circle" : "spiral prepare";
			if (this._state === "spiral prepare") {
				this._ticks = 1;
				this._increment = false;
			}
		} else if (this._state === "spiral prepare") {
			let origin = Vector.fromPolar(this._centerAngle, (this._radius / 2) * this._ticks);
			pos = this._center + origin + Vector.fromPolar(this._outerAngle, this._radius / (4 - this._ticks * 2));
			this._state = this._ticks > 0 ? "spiral prepare" : "spiral";
		} else if (this._state === "spiral") {
			pos = this._center + Vector.fromPolar(this._outerAngle, this._radius - this._ticks * this._radius * 3 / 4);
			this._state = this._centerAngle > 2 * NUM_OF_REVOLUTIONS * 2 * Math.PI ? "double circle prepare" : "spiral";
			if (this._state === "double circle prepare") {
				this._centerAngle = geom.normalizeAnglePositive(this._centerAngle);
				this._ticks = 0;
				this._increment = true;
			}
		} else if (this._state === "double circle prepare") {
			let origin = Vector.fromPolar(this._centerAngle, (this._radius / 2) * this._ticks);
			pos = this._center + origin + Vector.fromPolar(this._outerAngle, this._radius / (4 - this._ticks * 2));
			this._state = this._ticks > 1 ? "double circle" : "double circle prepare";
		}

		if (this._increment) {
			this._ticks = this._ticks + 0.002;
			this._increment = this._ticks < 1;
		} else {
			this._ticks = this._ticks - 0.002;
			this._increment = this._ticks < 0;
		}
		this._robot.path.clearObstacles();
		let endSpeed = new Vector(0, 0);
		let dir = (<Position> pos - this._robot.pos).angle();
		PathHelper.setDefaultObstaclesByTable(this._robot.path, this._robot, this._obstacleTable);
		this._robot.trajectory.update(ToTarget, pos!, dir, 1, endSpeed);
	}

}
